docs: adds interceptor testing

JsantosDK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Fixes #54390

JeanMeche commented on adev/src/content/guide/http/testing.md:
```diff
@@ -138,3 +138,73 @@ req.error(new ProgressEvent('network error!'));
 // Assert that the application successfully handled the network error.
 </docs-code>
+
+## Testing a Interceptor
```
```suggestion
## Testing an Interceptor
```

JsantosDK replied:
Fixed.

JeanMeche commented on adev/src/content/guide/http/testing.md:
```diff
@@ -138,3 +138,73 @@ req.error(new ProgressEvent('network error!'));
+expect(req.request.headers.get('Authorization')).toEqual('X-Authentication-Token', service.getAuthToken())
+</docs-code>
+
+A similar interceptor could be implement with DI.
```
```suggestion
A similar interceptor could be implemented with injected dependencies.
```

JsantosDK replied:
Fixed.

JeanMeche commented on adev/src/content/guide/http/testing.md:
```diff
@@ -138,3 +138,73 @@ req.error(new ProgressEvent('network error!'));
+  imports: [HttpClientTestingModule],
+  providers: [
+    // Due to the Interceptor now being an Injectable, we need to pass it in the providers array
+    AuthInterceptor,
```
This isn't necessary because you're using `inject()`

JsantosDK replied:
Updated `TestBed` configuration to:
```ts
TestBed.configureTestingModule({
  providers: [
    AuthService,
    provideHttpClient(withInterceptorsFromDi()),
    provideHttpClientTesting(), 
    // We rely on the HTTP_INTERCEPTORS token to register the AuthInterceptor as an HttpInterceptor
    { provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true },
  ],
});
```

Was also using the deprecated `HttpClientTestingModule` instead of the `providing`

JeanMeche reviewed (changes requested):
More generaly I think this guide should illustrate testing a `HttpInterceptorFn` since this is the recommended implementation for Interceptors.

JsantosDK commented:
> More generaly I think this guide should illustrate testing a `HttpInterceptorFn` since this is the recommended implementation for Interceptors.

In general, I agree with this, hence why I added the example for the functional interceptor first.

However, I was not sure how to proceed in this case, given that Angular still supports interceptors from DI.
Because of this, not having any details on how to test these interceptors seems a bit wrong, hence why I added them.

Let me know if these should be removed, or changed in some way.

JeanMeche commented on adev/src/content/guide/http/testing.md:
```diff
@@ -138,3 +138,73 @@ req.error(new ProgressEvent('network error!'));
+expect(req.request.headers.get('X-Authentication-Token')).toEqual(service.getAuthToken());
+</docs-code>
+
+A similar interceptor could be implemented with injected dependencies.
```
This work the same for `HttpInterceptorFn`, could you make this also a `HttpInterceptorFn` example ? Thank you.

JsantosDK replied:
The `HttpInterceptorFn` in lines 150-157 already uses the injected dependencies (it injects the `AuthService` in line 151. 
> A similar interceptor could be implemented with injected dependencies.

Perhaps this is not clear a sentence?
It should refer to an interceptor example that uses DI / is class based.
Otherwise, I'm sorry but I didn't understand what you mean @JeanMeche

JeanMeche replied:
`inject` is in fact DI, so function using have injected dependencies, so yes we should rephrase that. 
You could call them class based interceptors.

JsantosDK replied:
Changed 
> A similar interceptor could be implemented with injected dependencies.
to
> A similar interceptor could be implemented with class based interceptors:

This is more explicit and reduces the chances for confusion.

JsantosDK commented:
@JeanMeche any word on this?
Is there anything blocking it?

JeanMeche reviewed (approved):
LGTM, at least it a good start on that topic

JeanMeche commented:
We'll need you to drop that merge commit, the change needs to be a single commit to be merged.

JeanMeche commented:
We'll need you to drop that merge commit, the change needs to be a single commit to be merged.

JsantosDK commented:
@JeanMeche could you let me know which commit you want me to drop?
Or if I should drop this PR and open a new one?

JsantosDK committed:
docs: adds interceptor testing

JsantosDK commented:
@JeanMeche is there anything pending on my part, or is it just tag updating and merging?

thePunderWoman commented:
This PR was merged into the repository by commit 6674130e803b4829130c914ea883c75300a00313.

The changes were merged into the following branches: main, 18.2.x, 19.0.x
